Add Letter spacing and Letter case to Global Styles

noisysocks opened the pull request:
## What?
Two changes:

- Makes _Letter spacing_ (`LetterSpacingControl`) appear in Global Styles → Typography → Headings.
- Adds _Letter case_ (`TextTransformControl`) to Global Styles, and makes it appear in Global Styles → Typography → Headings.

This branch includes the commits in https://github.com/WordPress/gutenberg/pull/43328. Once that PR lands I'll rebase this one. In the meantime, only focus on dcfb5ee1d531077830de31c1583703155556c43b and c133c2f26bba8db61e75168049d3b6a226477f46.

## Why?
So that the Typography panel in Global Styles looks closer to what was designed in Figma. See https://github.com/WordPress/gutenberg/issues/34345#issuecomment-1217390408.

In particular, I'm implementing this suggestion from @mtias https://github.com/WordPress/gutenberg/issues/34345#issuecomment-1217479784:

> They should be available for headings in the global context

## How?
_Letter spacing_ was already set up in Global Styles, just not enabled for headings at the global level. To fix this I added a special condition to `useHasLetterSpacingControl`.

_Letter case_ was not in Global Styles at all. The control already exists so it's just a matter of rendering them in `TypographyPanel` in a manner similar to what we do for _Letter spacing_.

## Testing Instructions
1. Go to Editor → Global Styles → Typography.
2. Check that _Letter spacing_ and _Letter case_ appears in the Heading section but no other sections, and that modifying them works.
2. Go to Global Styles → Blocks.
3. Check that _Letter spacing_ and _Letter case_ appears in the Typography panel for blocks that support these controls, and that modifying them works.

## Screenshots or screencast
https://user-images.githubusercontent.com/612155/185303829-85ff0e19-c410-44f7-afa0-7fecbd0e937d.mp4

## To do
- [ ] Rebase when https://github.com/WordPress/gutenberg/pull/43328 merges.
- [ ] Come up with solution for unsetting _Letter case_.

noisysocks committed:
Show Letter Spacing in Global Styles -> Typography -> Headings

noisysocks committed:
Add Letter Case to Global Styles

aaronrobertshaw reviewed (commented):
Thanks for picking this one up @noisysocks 👍 

Nice work, everything tests as advertised. 

✅ Letter Spacing / Case controls only appear in Heading section under Global Styles > Typography
✅ Letter Spacing / Case controls only appeared for blocks with corresponding support

> Come up with solution for unsetting Letter case.

I was about to note this as a problem until I saw your todo item. I've voiced my concerns over the use of a `ToggleGroupControl` before. I'd like to see a solution to this problem before the switch is made.

We might get away with using the `ToggleGroupControl` in the block editor given it's within a `ToolsPanel` that provides us with a means of resetting the control to an indeterminate state. In the site editor's global styles, we only have the "Reset to defaults" nuclear option.

noisysocks commented:
> We might get away with using the `ToggleGroupControl` in the block editor given it's within a `ToolsPanel` that provides us with a means of resetting the control to an indeterminate state. In the site editor's global styles, we only have the "Reset to defaults" nuclear option.

Yeah agree. A design solution is needed. There's some discussion on this in https://github.com/WordPress/gutenberg/issues/36735.

noisysocks committed:
Show 'None' Letter Case option when in Global Styles

noisysocks commented:
I rebased this to include https://github.com/WordPress/gutenberg/pull/43328 and added a _None_ option to the _Letter Case_ control when it's shown in Global Styles as per https://github.com/WordPress/gutenberg/issues/36735#issuecomment-1237731168. The block inspector continues to use the Tools Panel flow.

https://user-images.githubusercontent.com/612155/189290200-c9f7bc5a-0e5b-44cf-b110-7479ec1de6de.mp4

What do you think @aaronrobertshaw @jasmussen?

Something I'm unsure about is whether _None_ is the right copy. Perhaps _Default_ would be better as then it implies that the value is inherited from the next level up in the Global Styles hierarchy?

jasmussen commented:
Thanks for the work! Global styles before, had no text letter case option:
![gs before](https://user-images.githubusercontent.com/1204802/189304186-f793721d-cb09-4a6a-974d-cca9acee85ec.gif)

Now it does, cool, and with a "none" setting:

![gs after](https://user-images.githubusercontent.com/1204802/189304224-723ad443-156a-437f-bfac-e11b91453e19.gif)

Letter case in the post editor before:

![editor before](https://user-images.githubusercontent.com/1204802/189304244-e2ec167d-cb08-4497-bb9c-00ed18de9a13.gif)

... and after:

![editor after](https://user-images.githubusercontent.com/1204802/189304270-511d372b-281a-4709-aa79-982ff19e7d0c.gif)

A few things:

* The segmented control should not be bordered, either in the editor or global styles. This is something I know we've discussed as an option (CC: @mirka), but for now let's try without it, i.e. as it ships today. It still has a focus style, it's just the resting state that shouldn't be bordered.
* At the moment, it looks like the "None" option is chosen by default, which isn't entirely accurate. There's both _unconfigured_, and _explicitly unset_, and we need to support both. So unless you've explicitly chosen _none_, I would expect no letter case option to be chosen.
* It'd be nice to also have the "None" option in the post editor.

I think the minus works good enough to try this out, and "None" is an okay label since it technically matches CSS. It also works to answer the question: "Text decoration?" "None!" 😅 

Is this useful?

noisysocks commented:
> * The segmented control should not be bordered, either in the editor or global styles. This is something I know we've discussed as an option (CC: @mirka), but for now let's try without it, i.e. as it ships today. It still has a focus style, it's just the resting state that shouldn't be bordered.

I saw that there was a border in the Figma but yep happy to go back to borderless 👍

> * At the moment, it looks like the "None" option is chosen by default, which isn't entirely accurate. There's both _unconfigured_, and _explicitly unset_, and we need to support both. So unless you've explicitly chosen _none_, I would expect no letter case option to be chosen.
> * It'd be nice to also have the "None" option in the post editor.

How would you switch from one of the select options to _unconfigured_? We don't have a Tools Panel in Global Styles, and the control is a radio meaning that you can't click a selected option to deselect it.

What I'm hearing is that https://github.com/WordPress/gutenberg/pull/43328 was the wrong approach and that we need to go back to the drawing board and figure out a way to implement click-to-deselect in an accessible way. That way we can distinguish between _unconfigured_ and _explicitly unset_ in Global Styles. I'll look into it.

noisysocks commented:
OK. I opened https://github.com/WordPress/gutenberg/pull/44067 as an alternative to this PR. I believe I implemented the UX that you're after @jasmussen where there is a distinction between _unconfigured_, and _explicitly unset_. I chose to implement this by partially reverting https://github.com/WordPress/gutenberg/pull/43328 and did my best to explain why I think this is okay to do in the description cc. @mirka.

noisysocks commented:
Closing in favour of https://github.com/WordPress/gutenberg/pull/44067.
